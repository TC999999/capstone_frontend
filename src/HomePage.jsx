import { useState, useEffect, useContext } from "react";
import UserContext from "./UserContext";
import marketAPI from "../api";
import FrontItemList from "./items/FrontItems.jsx";
import "./styles/HomePage.css";

const Home = () => {
  const { user } = useContext(UserContext);
  const [reccomendedItems, setReccomendedItems] = useState([]);
  const [itemsInLocation, setItemsInLocation] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  //gets both reccomended items based on type and location
  const getFrontPageItems = async () => {
    try {
      setIsLoading(true);
      let reccomendedItems = await marketAPI.getReccomendedItems(user.username);
      let itemsInLocation = await marketAPI.getItemsInLocation(user.username);
      setReccomendedItems(reccomendedItems);
      setItemsInLocation(itemsInLocation);
      setIsLoading(false);
    } catch (err) {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    setIsLoading(true);
    if (user) {
      getFrontPageItems();
    } else {
      setIsLoading(false);
    }
  }, [user]);

  if (isLoading) {
    return <p className="loading-message">Loading...</p>;
  }

  return (
    <div className="home-div">
      <h1>Welcome to The Worldwide Garage Sale!</h1>
      {!user && <h2>Please log in for full experience!</h2>}
      {user && (
        <div className="front-page-user">
          <h2>Reccomended Items for {user.username}</h2>
          <div className="front-page-items">
            <div className="reccomended-items-list-div">
              <h2>Items For You!</h2>
              <FrontItemList items={reccomendedItems} />
            </div>
            <div className="items-near-location-list-div">
              <h2>Items Near You!</h2>
              <FrontItemList items={itemsInLocation} />
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default Home;
