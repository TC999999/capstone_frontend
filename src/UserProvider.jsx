import UserContext from "./UserContext.js";
import { useState, useEffect } from "react";
import marketAPI from "../api.js";

const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  //sets context for user for entire app
  const currentUser = (user) => {
    setUser(user);
  };

  // //On initial render, get user token from local storage, gets the current user from the api, and set as context
  useEffect(() => {
    async function getToken() {
      let token = localStorage.getItem("market-token");
      if (token) {
        let userInfo = await marketAPI.getCurrentUser();
        currentUser(userInfo);
      }
    }
    getToken();
  }, []);

  //These functions and user state will be used in context
  //Set the token in local storage, retrives the current user from the api, and sets the user as context on login
  const logIn = async (username, token) => {
    localStorage.setItem("market-token", token);
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  //remove token and user from local storage and set user context to null on logout
  const logOut = () => {
    localStorage.removeItem("market-token");
    currentUser(null);
  };

  //update current user in local storage and context
  const updateUser = async (username) => {
    let userInfo = await marketAPI.getUserInfo(username);
    currentUser(userInfo);
  };

  return (
    <UserContext.Provider
      value={{ user, currentUser, logIn, logOut, updateUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
