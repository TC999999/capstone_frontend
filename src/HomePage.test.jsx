import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "./UserProvider.jsx";
import HomePage from "./HomePage.jsx";

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <UserProvider>
        <HomePage />
      </UserProvider>
    </MemoryRouter>
  );
});

it("should match snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <HomePage />
      </UserProvider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("should have a header", function () {
  const { getByText } = render(
    <MemoryRouter>
      <UserProvider>
        <HomePage />
      </UserProvider>
    </MemoryRouter>
  );

  const header = getByText("Welcome to The Worldwide Garage Sale!");
  expect(header).toBeInTheDocument;
});
