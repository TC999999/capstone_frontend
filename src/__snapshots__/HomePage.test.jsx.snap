// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="home-div"
  >
    <h1>
      Welcome to The Worldwide Garage Sale!
    </h1>
    <h2>
      Please log in for full experience!
    </h2>
  </div>
</DocumentFragment>
`;
